File: problems/[E]1.Two-Sum.kt

/**
 * Runtime: 164 ms, faster than 98.92% of Kotlin online submissions for Two Sum.
 * Memory Usage: 36.9 MB, less than 56.61% of Kotlin online submissions for Two Sum.
 */
class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val numsSize = nums.size
        val addendNumbersSet = HashSet<Int>(numsSize)
        var secondNumberIndex: Int = -1
        var secondNumber: Int = -1
        for (i in 0 until numsSize) {
            if (nums[i] !in addendNumbersSet) {
                addendNumbersSet.add(target - nums[i])
            } else {
                secondNumberIndex = i
                secondNumber = nums[i]
                break
            }
        }
        val firstNumber = target - secondNumber
        var firstNumberIndex = target - secondNumber
        for (i in 0 .. secondNumberIndex) {
            if (nums[i] == firstNumber) {
                firstNumberIndex = i
                break
            }
        }
        return intArrayOf(firstNumberIndex, secondNumberIndex)
    }
}

File: python/[E]1.Two-Sum.py
# Runtime: 44 ms, faster than 85.74% of Python3 online submissions for Two Sum.
# Memory Usage: 14.2 MB, less than 98.60% of Python3 online submissions for Two Sum.
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        adds = {}
        for i in range(len(nums)):
            number = nums[i]
            add = target - number
            if add in adds:
                return [adds.get(add), i]
            else:
                adds[number] = i
        return []
        
